### ISSUE TITLE
Question: how to call script function from Object

### ISSUE THREAD

**Issue by j-o** [cginternals/cppexpose#51]
Suppose I have an instance of a `class MyObject : public cppexpose::Object` that is exposed somewhere in the script context. Now I add/set a JS function via the script console: 
```a.b.c.myObject.onProcess = function(data) {...};```

How can I call this function from within the C++-implementation of `MyObject` without knowing its fully qualified name (i.e., the `a.b.c` part)? E.g.:
```
void MyObject::process()
{
    Variant result = ... // call onProcess here
}
```

Second part of this question: When issuing such a call, how can I directly pass an `Object` as argument, again without knowing its fully qualified name in the scripting context?

**Comment by sbusch42** [comment]
When a cppexpose::Object is exposed into a scripting context, a representation of that object is created in the scripting context, but it is not the object itself. Therefore, properties which are added to the wrapped object will not be accessible from C++, at least I don't see a way to do that.

So, instead of
`a.b.c.myObject.onProcess = function(data) {...};`

I would suggest this:
`a.b.c.myObject.onProcess(function(data) {...});`

And on the C++ side:
```
class MyObject : public cppexpose::Object
{
public:
  MyObject() {
    addFunction("onProcess", this, MyObject::onProcess);
  }
  void onProcess(const Variant & func) {
        if (func.hasType<Function>())
        {
            m_onProcess = func.value<Function>();
        }
  }
  void process() {
    std::vector<Variant> args;
    // args.push_back(...);
    Variant res = m_onProcess.call(args);
  }

  cppexpose::Function m_onProcess;
}
```

I have to think about the second part.

**Comment by j-o** [comment]
Thank you, that seems a good and simpler to implement alternative.

I was experimenting with `Property<Function>`, which requires a little work on the scripting backends to make these callable from JS. This lead me to two further questions:
1. The whole `Function`, `AbstractFunction`, `Method`, etc. construct looks a lot like `std::function<Variant(VariantArray)>`, why duplicate this functionality?
2. Would it be a valid idea to drop the special treatment of functions for an `Object` and instead make them "first-class" properties as mentioned above (including dynamic functions)?

**Comment by cgcostume** [comment]
can this be closed?

**Comment by sbusch42** [comment]
Actually, I implemented cppexpose 2.0 with these comments in mind. In cppexpose 2.0, functions are indeed stored as Var<Function>, instead of having a separate list as in cppexpose 1.0. As for the other question, I'm not totally sure how to implement this with std::function only, as we need the interface to call functions using variants. So maybe we still need Function, but the backend-classes could be replaced. I would appreciate suggestions on how to implement this :)
